File: php/15_binarysearch/binarysearch.php
<?php
/**
 * https://github.com/wangzheng0822/algo/blob/master/php/15_binary/binary.php
 * Created by PhpStorm.
 * User: duanfuhao
 * Date: 2019/9/26
 * Time: 14:59
 */
function binarysearch($arr,$val) {
    $len = count($arr);
    if ($len <= 1) return $arr;
    $low = 0;
    $high = $len - 1;
    $mid = (int)($low + ($high - $low) / 2);
    while ($low <= $high) {
        if ($arr[$mid] < $val) {
            $low = $mid + 1;
        } elseif ($arr[$mid] > $val) {
            $high = $mid - 1;
        } else {
            return $mid;
        }
    }
    return -1;
}

function bsearch($arr,$val) {
    return bsearch_rec($arr,$val,0,count($arr) - 1);
}

function bsearch_rec($arr,$val,$l,$h) {
    if ($l > $h) return -1;
    $mid = int($l + ($h - $l) / 2);
    if ($arr[$mid] < $val) {
        bsearch_rec($arr,$val,$mid + 1,$h);
    } elseif ($arr[$mid] > $val) {
        bsearch_rec($arr,$val,$l,$mid - 1);
    } else {
        return $mid;
    }
}

/**
 * 二分法求平方根,精确到小数点后6位
 * @param $val
 * @param $double
 * @author duanfuhao@smzdm.com
 * @date 2019/9/30
 */
function square($val, $double) {
    if ($val < 0) { //必须是正数
        return false;
    }
    $low = 0;
    $up = $val;
    if ($val > 0 && $val < 1) {
        $low = $val;
        $up = 1;
    }
    $mid = $low + ($up - $low) / 2;
    while ($up - $low > $double) {
        if ($mid * $mid > $val) {
            $up = $mid;
        } elseif ($mid * $mid < $val) {
            $low = $mid;
        } else {
            return $mid;
        }
        $mid = $low + ($up - $low) / 2;
    }
    return $mid;
}

var_dump(square(9.9, 0.000001));

File: go/src/15_binarysearch/BinarySearch.go
package _15_binarysearch

//二分查找循环
func BinarySearch(arr []int, val int) int {
	len := len(arr)
	if len == 0 {
		return -1
	}
	l := 0
	h := len - 1
	for l <= h {
		mid := l + (h-l)/2
		if arr[mid] > val { //查找值小于中间值，在中间值左边查找
			h = mid - 1
		} else if arr[mid] < val { //查找值大于中间值，在中间值右边查找
			l = mid + 1
		} else {
			return mid
		}
	}
	return -1
}

//二分查找递归
func BinarySearchRecursive(arr []int, val int) int {
	len := len(arr)
	if len == 0 {
		return -1
	}
	return bs(arr, 0, len-1, val)
}

func bs(arr []int, l, h, val int) int {
	if l > h {
		return -1
	}
	mid := l + (h-l)/2
	if arr[mid] == val {
		return mid
	} else if arr[mid] > val {
		return bs(arr, l, mid-1, val)
	} else {
		return bs(arr, mid+1, h, val)
	}
}

//查找第一个给定值的元素
func BinarySearchFirst(arr []int, val int) int {
	len := len(arr)
	if len == 0 {
		return -1
	}
	l := 0
	h := len - 1
	for l <= h {
		mid := l + (h-l)/2
		if arr[mid] > val { //查找值小于中间值，在中间值左边查找
			h = mid - 1
		} else if arr[mid] < val { //查找值大于中间值，在中间值右边查找
			l = mid + 1
		} else {
			if mid == 0 || arr[mid-1] != val {
				return mid
			} else {
				h = mid - 1
			}
		}
	}
	return -1
}

//查找最后一个给定值的元素
func BinarySearchLast(arr []int, val int) int {
	len := len(arr)
	if len == 0 {
		return -1
	}
	l := 0
	h := len - 1
	for l <= h {
		mid := l + (h-l)/2
		if arr[mid] > val { //查找值小于中间值，在中间值左边查找
			h = mid - 1
		} else if arr[mid] < val { //查找值大于中间值，在中间值右边查找
			l = mid + 1
		} else {
			if mid == h || arr[mid+1] != val {
				return mid
			} else {
				l = mid + 1
			}
		}
	}
	return -1
}

//查找第一个大于等于给定值的元素
func BinarySearchGT(arr []int, val int) int {
	len := len(arr)
	if len == 0 {
		return -1
	}
	l := 0
	h := len - 1
	for l <= h {
		mid := l + (h-l)/2
		if arr[mid] > val { //查找值小于中间值，在中间值左边查找
			h = mid - 1
		} else if arr[mid] < val { //查找值大于中间值，在中间值右边查找
			l = mid + 1
		} else {
			if mid != len-1 && a[mid+1] > val {
				return mid + 1
			} else {
				l = mid + 1
			}
		}
	}
	return -1
}

//查找最后一个小于等于给定值的元素
func BinarySearchLT(arr []int, val int) int {
	len := len(arr)
	if len == 0 {
		return -1
	}
	l := 0
	h := len - 1
	for l <= h {
		mid := l + (h-l)/2
		if arr[mid] > val { //查找值小于中间值，在中间值左边查找
			h = mid - 1
		} else if arr[mid] < val { //查找值大于中间值，在中间值右边查找
			l = mid + 1
		} else {
			if mid == 0 && a[mid-1] < val {
				return mid - 1
			} else {
				h = mid - 1
			}
		}
	}
	return -1
}
